Assets/Scripts/Enemies/BaseEnemyAI.cs

```diff
@@ -0,0 +1,136 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class BaseEnemyAI : MonoBehaviour
+{
+    public int currentHealth;
+
+    [SerializeField] Animator animator;
+    [SerializeField] MovementOverride movement;
+    [SerializeField] Rigidbody2D rb;
+    [SerializeField] Resource health;
+    [SerializeField] AttackData meleeAttackData;
+    [SerializeField] AttackData projectileAttackData;
+
+    [SerializeField] float detectionRadius;
+    [SerializeField] float meleeRange;
+    [SerializeField] Transform projectileSpawnPoint;
+
+    PlayerMovementController player;
+    PlayerCombat playerCombat;
+
+    Vector2 velocity;
+    bool isFacingRight = true;
+    float horizontalMove = 0.0f;
+    float maxSpeedChange;
+    float nextMeleeTime = 0f;
+    float nextProjectileTime = 0f;
+
+    // Values used for animations:
+
+    [SerializeField] bool isFiringProjectile = false;
+    [SerializeField] bool isAttacking = false;
+
+    void FindPlayer()
+    {
+        if (player == null)
+        {
+            player = GameObject.FindWithTag(Constants.PLAYER_TAG).GetComponent<PlayerMovementController>();
+        }
+        if (playerCombat == null)
+        {
+            playerCombat = GameObject.FindWithTag(Constants.PLAYER_TAG).GetComponent<PlayerCombat>();
+        }
+    }
+
+    void Update()
+    {
+        FindPlayer();
+        float distanceToPlayer = Mathf.Abs(player.transform.position.x - transform.position.x);
+        if (distanceToPlayer < detectionRadius && distanceToPlayer > meleeRange && !isAttacking && !isFiringProjectile)
+        {
+            Debug.Log("approaching player");
+            horizontalMove = (player.transform.position - transform.position).normalized.x;
+            if (horizontalMove > 0 && !isFacingRight)
+            {
+                Flip();
+            }
+            else if (horizontalMove < 0 && isFacingRight)
+            {
+                Flip();
+            }
+            if (Time.time >= nextProjectileTime)
+            {
+                Debug.Log("fire");
+                nextProjectileTime = Time.time + projectileAttackData.duration;
+                animator.SetTrigger("ProjectileAttack");
+            }
+        }
+        else
+        {
+            if (distanceToPlayer < meleeRange && Time.time >= nextMeleeTime && !isFiringProjectile)
+            {
+                Debug.Log("attack");
+                nextMeleeTime = Time.time + meleeAttackData.duration;
+                animator.SetTrigger("MeleeAttack");
+            }
+        }
+    }
+
+    void FixedUpdate()
+    {
+        if (isAttacking || isFiringProjectile)
+        {
+            horizontalMove = 0f;
+        }
```

Comment:
Very minor - I think that it would be better to place this in Update just after the initial move calc on line 54: https://github.com/MitchellCoon/MetroidvaniaGameJamFeb23/pull/39/files#diff-764d5591104aba6d194a3b5c59a458336dce7c532205c77d04f979fb22ba67faR54

However this will overwrite `horizontalMove` each time and is perfectly fine. But having `horizontalMove` be updated in a single place will make it easier to avoid bugs IMO

Reply:
It's a little complicated, but the isFiringProjectile and isAttacking values are set when the animations start, by which point the horizontal move would have already been calculated, and because of the other conditions, it won't be updated again until after the animation ends, so if we moved this code to after line 54, it wouldn't do anything. I would like to eventually update this so that the enemy changes direction to face the player before attacking based on the player's location relative to the enemy instead of the movement direction, but long story short, I'm leaving this as is because otherwise it won't work